File: app/src/main/java/yangFenTuoZi/runner/plus/ui/dialog/BaseDialogBuilder.java
package yangFenTuoZi.runner.plus.ui.dialog;

import android.content.DialogInterface;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;
import androidx.appcompat.app.AlertDialog;

import com.google.android.material.dialog.MaterialAlertDialogBuilder;

import yangFenTuoZi.runner.plus.ui.activity.BaseActivity;

public class BaseDialogBuilder extends MaterialAlertDialogBuilder {
    private final BaseActivity mBaseActivity;
    private AlertDialog mAlertDialog;
    private DialogInterface.OnDismissListener mOnDismissListener;

    public static class DialogShowException extends Exception {
    }

    public BaseDialogBuilder(@NonNull BaseActivity context) throws DialogShowException {
        super(context);
        mBaseActivity = context;
        if (mBaseActivity.isDialogShow) throw new DialogShowException();
        mBaseActivity.isDialogShow = true;
        super.setOnDismissListener(dialogInterface -> {
            mBaseActivity.isDialogShow = false;
            if (mOnDismissListener != null)
                mOnDismissListener.onDismiss(dialogInterface);
        });
    }

    public AlertDialog getAlertDialog() {
        return mAlertDialog;
    }

    @NonNull
    @Override
    public AlertDialog create() {
        return mAlertDialog = super.create();
    }

    @NonNull
    @Override
    public MaterialAlertDialogBuilder setOnDismissListener(@Nullable DialogInterface.OnDismissListener onDismissListener) {
        mOnDismissListener = onDismissListener;
        return this;
    }

    public void runOnUiThread(Runnable action) {
        mBaseActivity.runOnUiThread(action);
    }

    @NonNull
    public final String getString(@StringRes int resId, Object... formatArgs) {
        return mBaseActivity.getString(resId, formatArgs);
    }

    @NonNull
    public final String getString(@StringRes int resId) {
        return mBaseActivity.getString(resId);
    }
}

File: app/src/main/java/yangfentuozi/runner/base/BaseDialogBuilder.kt
package yangfentuozi.runner.base

import android.content.DialogInterface
import androidx.annotation.StringRes
import androidx.appcompat.app.AlertDialog
import com.google.android.material.dialog.MaterialAlertDialogBuilder

open class BaseDialogBuilder @Throws(DialogShowException::class) constructor(context: BaseActivity) : MaterialAlertDialogBuilder(context) {
    class DialogShowException : Exception()

    private val mBaseActivity: BaseActivity = context
    private var mAlertDialog: AlertDialog? = null
    private var mOnDismissListener: DialogInterface.OnDismissListener? = null

    init {
        if (mBaseActivity.isDialogShow) throw DialogShowException()
        mBaseActivity.isDialogShow = true
        super.setOnDismissListener { dialogInterface ->
            mBaseActivity.isDialogShow = false
            mOnDismissListener?.onDismiss(dialogInterface)
        }
    }

    fun getAlertDialog(): AlertDialog? = mAlertDialog

    
    override fun create(): AlertDialog {
        return super.create().also { mAlertDialog = it }
    }

    
    override fun setOnDismissListener(onDismissListener: DialogInterface.OnDismissListener?): MaterialAlertDialogBuilder {
        mOnDismissListener = onDismissListener
        return this
    }

    fun runOnUiThread(action: Runnable) {
        mBaseActivity.runOnUiThread(action)
    }

    
    fun getString(@StringRes resId: Int, vararg formatArgs: Any?): String {
        return mBaseActivity.getString(resId, *formatArgs)
    }

    
    fun getString(@StringRes resId: Int): String {
        return mBaseActivity.getString(resId)
    }
}